### PR TITLE
CodeGenerator plugin and Java code generation.

### PR DESCRIPTION
**Pull request by jdochez**

### CODE REVIEW

File: code/codegen/build.gradle.kts [hunk 1]
```diff
@@ -28,6 +29,19 @@ tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().configureEach
     }
 }
 
+// publish to maven local, it is where the init script is expecting it.
+publishing {
+    publications {
+        create<MavenPublication>("maven") {
+            groupId = "com.android.gradle.replicator"
+            artifactId = "codegen"
+            version = "0.1"
```
Comment (ducrohet) [inline code/codegen/build.gradle.kts:38]: you should use `Versions.pluginVersion`

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/ImportClassPicker.kt [hunk 2]
```diff
@@ -0,0 +1,172 @@
+package com.android.gradle.replicator.codegen
```
Comment (ducrohet) [inline code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/ImportClassPicker.kt:1]: missing header

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/gradlegen/groovy/GradleJavaBuildFileGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.android.gradle.replicator.codegen.gradlegen.groovy
```
Comment (ducrohet) [inline code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/gradlegen/groovy/GradleJavaBuildFileGenerator.kt:1]: missing header

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/java/JavaClassGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
+package com.android.gradle.replicator.codegen.java
```
Comment (ducrohet) [inline code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/java/JavaClassGenerator.kt:1]: missing header

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/kotlin/KotlinClassGenerator.kt [hunk 5]
```diff
@@ -43,6 +45,7 @@ class KotlinClassGenerator(
             listener.
             classDeclaration(name, CodeGenerationListener.Type.CLASS, listOf(), printer)
         }
+//        printer.printlnIndented("@kotlin.OptIn(kotlin.ExperimentalStdlibApi::class, kotlin.ExperimentalUnsignedTypes::class, kotlin.time.ExperimentalTime::class)")
```
Comment (ducrohet) [inline code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/kotlin/KotlinClassGenerator.kt:48]: remove or comment?

File: code/plugin/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+plugins {
```
Comment (ducrohet) [inline code/plugin/build.gradle.kts:1]: missing header.

File: code/plugin/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    implementation(project(":code:codegen"))
+}
+
+// publish to maven local, it is where the init script is expecting it.
+publishing {
+    publications {
+        create<MavenPublication>("maven") {
+            groupId = "com.android.gradle.replicator"
+            artifactId = "codegen-plugin"
+            version = "0.1"
```
Comment (ducrohet) [inline code/plugin/build.gradle.kts:23]: should use `Versions` as well.

File: code/plugin/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+            from(components["java"])
+        }
+    }
+}
+
+gradlePlugin {
+    val agpFunctionalTest by plugins.creating {
+        id = "com.android.gradle.replicator.codegen-plugin"
+        version = 0.1
```
Comment (ducrohet) [inline code/plugin/build.gradle.kts:33]: and here

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt [hunk 9]
```diff
@@ -0,0 +1,159 @@
+package com.android.gradle.replicator.codegen.plugin
```
Comment (ducrohet) [inline code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt:1]: missing header

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            task.codeGeneratedModuleImplClasspath.from(implProjectArtifacts.artifactFiles)
+            task.apiJarFiles.from(apiArtifacts.artifactFiles)
+            task.implJarFiles.from(androidApis, implArtifacts.artifactFiles)
+            task.paramsFile.set(project.layout.buildDirectory.file("test.params"))
+            // Randomizer values should be set during project replication along the number of java and kotlin files.
+            task.seed.set(Random.nextInt())
+            // make sure we depend on our dependencies built artifacts so we have access to their generated classes.
+            projectDependencies.forEach {
+                // hack : use Variant API when dealing with android.
+                task.dependsOn(if (project.tasks.findByName("$it:assembleDebug") != null) {
```
Comment (ducrohet) [inline code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt:139]: also, this force creation of the task. Sadly there's not a better API at the moment.

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateCode.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package com.android.gradle.replicator.codegen.plugin
```
Comment (ducrohet) [inline code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateCode.kt:1]: missing header

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateParamsTask.kt [hunk 12]
```diff
@@ -0,0 +1,58 @@
+package com.android.gradle.replicator.codegen.plugin
```
Comment (ducrohet) [inline code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateParamsTask.kt:1]: missing header

**Review by ducrohet — CHANGES_REQUESTED** [review]

**Comment by ducrohet** [comment]
There's a build failure. My guess is that you didn't resolve the conflict that was present in the previous patch. you should double check to see what's going on.

File: code/codegen/build.gradle.kts (hunk 1 above)
Comment (jdochez) [reply code/codegen/build.gradle.kts:38]: Done

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/ImportClassPicker.kt (hunk 2 above)
Comment (jdochez) [reply code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/ImportClassPicker.kt:1]: Done

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/gradlegen/groovy/GradleJavaBuildFileGenerator.kt (hunk 3 above)
Comment (jdochez) [reply code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/gradlegen/groovy/GradleJavaBuildFileGenerator.kt:1]: Done

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/java/JavaClassGenerator.kt (hunk 4 above)
Comment (jdochez) [reply code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/java/JavaClassGenerator.kt:1]: Done

File: code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/kotlin/KotlinClassGenerator.kt (hunk 5 above)
Comment (jdochez) [reply code/codegen/src/main/kotlin/com/android/gradle/replicator/codegen/kotlin/KotlinClassGenerator.kt:48]: Done

File: code/plugin/build.gradle.kts (hunk 8 above)
Comment (jdochez) [reply code/plugin/build.gradle.kts:33]: Done

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt (hunk 9 above)
Comment (jdochez) [reply code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt:1]: Done

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt (hunk 10 above)
Comment (jdochez) [reply code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/CodegenPlugin.kt:139]: I should be able to depend on AAR, i'll look into it.

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateCode.kt (hunk 11 above)
Comment (jdochez) [reply code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateCode.kt:1]: Done

File: code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateParamsTask.kt (hunk 12 above)
Comment (jdochez) [reply code/plugin/src/main/kotlin/com/android/gradle/replicator/codegen/plugin/GenerateParamsTask.kt:1]: Done

**Pull request closed without merging**